### PR TITLE
Rewrite main to use argparse and add additional options

### LINKED ISSUE #500: Allow to run with `pipx`
I don't know what it takes to make it work better with `pipx`, but it would be super nice if possible.

```console
$ docker run --rm -it ubuntu
$ apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends pipx
$ pipx install tinytuya
Note: Dependent package 'charset-normalizer' contains 1 apps
  - normalizer

No apps associated with package tinytuya. Try again with '--include-deps' to include apps of dependent packages, which are listed above. If you are attempting to
install a library, pipx should not be used. Consider using pip or a similar tool instead.
```

I am no Python expert. Maybe you just need to declare the tinytuya executable/binary in some manifest? Also, not sure about this "normalizer" thing.

### LINKED ISSUE #501: No Timeout or Error, Just Hangs
Problem reported:

> I’m using the library but when we run the status call it just holds there when no device is online, it doesn’t fail or error it just hangs there waiting for it? Is there a way for it to throw an error or just return “false” or something if a Tuya device isn’t present?

### PR DESCRIPTION
**Pull request by uzlonewolf**
I just need to add a few comments to explain some of the code and bang on it a bit to make sure I didn't forget to implement any of the options and I think it's done.

Also closes #500 

## v0.14.0

* Rewrite of main to use argparse and add additional options
* Add support for `pipx install tinytuya` as raised by @felipecrs in https://github.com/jasonacox/tinytuya/issues/500 allowing for easier CLI use.
* Updated docs to explain timeout as raised by @GamerPeggun in https://github.com/jasonacox/tinytuya/issues/501

Also closes #501

### CODE REVIEW

**Comment by jasonacox** [comment]
I like this. It includes my updates so I'm good pushing this one forward instead.

**Comment by jasonacox** [comment]
To test pipx from parent directory of tinytuya:

```bash
pipx install ./tinytuya
```

It required that I create a `pyproject.toml` to install dependencies in the build phase:

```toml
[build-system]
requires = ["setuptools", "colorama", "requests", "cryptography"]
```

**Comment by jasonacox** [comment]
I was able to `pipx` install this PR and started some testing...

```bash
$ tinytuya 
TinyTuya Setup Wizard [1.14.0]

    Enter API Key from tuya.com: 
```

It looks like default is wizard. Do we want that or the usage map?

```bash
tinytuya -h  
usage: python -m tinytuya [-h] [-debug] {wizard,scan,devices,snapshot,json} ...

TinyTuya [1.14.0]

positional arguments:
  {wizard,scan,devices,snapshot,json}

options:
  -h, --help            show this help message and exit
  -debug, -d            Enable debug messages
```

This was the old help - I think we can probably get some of the descriptors back in there:

```
TinyTuya [1.13.2]

Usage:

    python -m tinytuya <command> [<max_time>] [-debug] [-nocolor] [-force [192.168.0.0/24 192.168.1.0/24 ...]] [-h]

      wizard         Launch Setup Wizard to get Tuya Local KEYs.
      scan           Scan local network for Tuya devices.
      devices        Scan all devices listed in devices.json file.
      snapshot       Scan devices listed in snapshot.json file.
      json           Scan devices listed in snapshot.json file [JSON].
      <max_time>     Maximum time to find Tuya devices [Default=18]
      -nocolor       Disable color text output.
      -force         Force network scan of device IP addresses based on format:
                     [net1/mask1 net2/mask2 ...] Auto-detects if none provided.
      -no-broadcasts Ignore broadcast packets when force scanning.
      -debug         Activate debug mode.
      -h             Show usage.
```

**Comment by uzlonewolf** [comment]
> It looks like default is wizard. Do we want that or the usage map?

I was just about to ask that.  Although I'm leaning towards displaying the help, I'm fine either way.  I made it auto-run the wizard since that's what it does now.  What do you think would be better?

>  I think we can probably get some of the descriptors back in there

The issue is how argparse handles subparsers.  I don't know of any good way to add them without having the command specified.  The list of options also changes depending on which command is given.

**Comment by jasonacox** [comment]
> Although I'm leaning towards displaying the help

Same!  Yes, I think the usage help is the right option. It shows the version and commands.

> The issue is how argparse handles subparsers. I don't know of any good way to add them without having the command specified. The list of options also changes depending on which command is given.

I know, it may break it from current use, but is probably the cleaner thing to do. We could have each command as a subparser.  This is what we did for pypowerwall:

https://github.com/jasonacox/pypowerwall/blob/110f43870d0c6fd0661c343c850e9f1dd504e1a0/pypowerwall/__main__.py#L34-L191

Gives:

```
$ python3 -m pypowerwall -h                                                            main
usage: PyPowerwall [-h] [-debug] {setup,fleetapi,scan,set,get,version} ...

PyPowerwall Module v0.9.1

options:
  -h, --help            show this help message and exit
  -debug                Enable debug output

commands (run <command> -h to see usage information):
  {setup,fleetapi,scan,set,get,version}
    setup               Setup Tesla Login for Cloud Mode access
    fleetapi            Setup Tesla FleetAPI for Cloud Mode access
    scan                Scan local network for Powerwall gateway
    set                 Set Powerwall Mode and Reserve Level
    get                 Get Powerwall Settings and Power Levels
    version             Print version information
```

**Comment by uzlonewolf** [comment]
> This is what we did for pypowerwall

Ah, that's what I wanted but wasn't able to figure out how to do.  Thanks for the example, I'll get that implemented.

**Comment by uzlonewolf** [comment]
```
$ tinytuya
usage: tinytuya [-h] [-debug] {wizard,scan,devices,snapshot,json} ...

TinyTuya [1.14.0]

options:
  -h, --help            show this help message and exit
  -debug, -d            Enable debug messages

commands (run <command> -h to see usage information):
  {wizard,scan,devices,snapshot,json}
    wizard              Launch Setup Wizard to get Device Local Keys
    scan                Scan local network for Tuya devices
    devices             Scan all devices listed in device-file
    snapshot            Scan devices listed in snapshot-file
    json                Scan devices listed in snapshot-file and display the result as JSON
```
```
$ python3 -m tinytuya
usage: python3 -m tinytuya [-h] [-debug] {wizard,scan,devices,snapshot,json} ...

TinyTuya [1.14.0]

options:
  -h, --help            show this help message and exit
  -debug, -d            Enable debug messages

commands (run <command> -h to see usage information):
  {wizard,scan,devices,snapshot,json}
    wizard              Launch Setup Wizard to get Device Local Keys
    scan                Scan local network for Tuya devices
    devices             Scan all devices listed in device-file
    snapshot            Scan devices listed in snapshot-file
    json                Scan devices listed in snapshot-file and display the result as JSON
```

**Comment by uzlonewolf** [comment]
I banged on it pretty good and only found 1 minor bug with wizard.  That's fixed and I made a couple other tweaks while I was at it.  So, I'm going to call this one good to go.

**Comment by uzlonewolf** [comment]
> 1. Maybe you should refactor README now to suggest using pipx instead of pip. (developers who understands and wants it to be used as a library will know they should install it with pip/poetry whatever).

No, pip would be the recommended way with pipx just being an option as the tinytuya CLI is rather pointless by itself.  The vast majority of users use tinytuya as a library to control devices using their own scripts.

> 2\. but the `--help` function shows `python -m tinytuya`, and I think that's unnecessary now, it can probably just be `tinytuya`

That actually changes depending on how it was called, as shown in the example I posted above.
```
$ tinytuya
usage: tinytuya [-h] [-debug] {wizard,scan,devices,snapshot,json} ...
```
```
$ python3 -m tinytuya
usage: python3 -m tinytuya [-h] [-debug] {wizard,scan,devices,snapshot,json} ...
```

**Comment by felipecrs** [comment]
> That actually changes depending on how it was called, as shown in the example I posted above.

I see. I was testing with `1.13.3b1`, maybe that's why.

**Comment by felipecrs** [comment]
> No, pip would be the recommended way with pipx just being an option as the tinytuya CLI is rather pointless by itself. The vast majority of users use tinytuya as a library to control devices using their own scripts.

Sure, it's your call. By the way I use tinytuya because of tuya-local. To find out the devices information to supply to it. I suppose several users also do the same.

> That actually changes depending on how it was called, as shown in the example I posted above.

I see. I was testing with `1.13.3b1`, maybe that's why.

**Comment by uzlonewolf** [comment]
> By the way I use tinytuya because of tuya-local. To find out the devices information to supply to it. I suppose several users also do the same.

Sure, you can use it for that, but if that's all you need it for there are other much easier ways of getting the keys (such as https://github.com/blakadder/tuya-uncover ).

**Comment by felipecrs** [comment]
Oh. That's cool. Thank you!

File: tinytuya/BulbDevice.py [hunk 1]
```diff
@@ -191,6 +191,9 @@ def _hexvalue_to_rgb(hexvalue, bulb="A"):
             r = int(rgb[0] * 255)
             g = int(rgb[1] * 255)
             b = int(rgb[2] * 255)
+        else:
+            # should we raise an error instead?
+            r = g = b = 0
```
[the remark this answers is not in the document]
Comment (jasonacox) [reply tinytuya/BulbDevice.py]: Before this conditional change, this call would error if the bulb type was not A or B. To keep that consistent but make it more informative, we now raise an error, e.g.

```python
raise ValueError(f"Unsupported bulb type {bulb} - unable to determine HSV values.")
```

**Review by jasonacox — APPROVED** [review]
LGTM

**Comment by jasonacox** [comment]
Great work on this @uzlonewolf - I love the move to argparse and the additional arguments. Thank you! 🙇  I tweaked the docs to represent this change and added a RELEASE note.

@felipecrs Thanks for your help and feedback!  While I agree with @uzlonewolf that this is primarily a library, the `pipx` CLI support is a very nice addition. I use the command line options quite a bit myself. 😁 I updated the README to highlight that and how to install and use. Again, thank you! 🙏

**Comment by felipecrs** [comment]
Awesome!

**Comment by jasonacox** [comment]
Released: https://pypi.org/project/tinytuya/1.14.0/

```bash
pipx install tinytuya
```

**Pull request merged**